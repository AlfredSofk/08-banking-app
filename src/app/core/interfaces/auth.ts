
interface IAuthUser {
    id: string;
    name: string;
}

export interface IAuthState {
    isAuthenticated: boolean;
    user: IAuthUser | null;
    token: string | null;
    loading: boolean;
    error: string | null;
}