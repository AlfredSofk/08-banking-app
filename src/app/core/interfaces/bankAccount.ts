
interface ITransaction {
    id: string;
    amount: number;
    type: 'DEPOSIT' | 'WITHDRAW' | 'TRANSFER' | 'PURCHASE';
    description: string;
    createdAt: Date;
}

interface IUserAccounts {
    number: string;
    amount: number;
}

export interface IAppState {
    saldoGlobal: number;
    transacciones: ITransaction[];
    username: string;
    userAccounts: IUserAccounts[];
    loading: boolean;
    error: string | null;
}