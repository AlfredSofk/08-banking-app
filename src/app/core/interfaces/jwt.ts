export interface IJWT {
    authorities: Array<string>
    exp : number
    iat: number
    jti: string
    sub: string
}