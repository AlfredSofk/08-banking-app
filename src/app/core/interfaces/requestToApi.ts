export interface IDinBodyGetDataClient {
    username: string;
}

export interface IRequestBodyWithDrawalATM {
    accountNumber: string;
    amount: number;
}

export interface IDinHeader {
    dispositivo: string;
    idioma: string;
    uuid: string;
    ip: string;
    horaTransaccion: string;
    llaveSimetrica: string;
    vectorInicializacion: string;
}


export interface IDinError {
    tipo: string;
    fecha: string;
    origen: string;
    codigo: string;
    codigoErrorProveedor: string;
    mensaje: string;
    detalle: string;
}

export interface IBodyLoginToken {
    username: string;
    password: string;
}

export interface IDinBodyLoginToken {
    dinHeader: IDinHeader;
    dinBody: IBodyLoginToken;
}

export interface IRequestMs {
    dinHeader: IDinHeader;
    dinError: IDinError;
}

export interface IRequestBodyGetDataClient {
    dinHeader: IDinHeader;
    dinBody: IDinBodyGetDataClient
}

export interface IFormBodyTransaction {
    accountNumber?: string;
    accountNumberClient?: string;
    accountNumberSender?: string;
    accountNumberReceiver?: string;
    typeBuys?: number;
    amount: number
}
export interface IRequestBodyTransaction {
    dinHeader: IDinHeader
    dinBody: IFormBodyTransaction

}

