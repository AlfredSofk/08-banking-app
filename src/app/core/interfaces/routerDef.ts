import  { ReactElement } from 'react';

// Tipo para las rutas hijas
interface RouteChild {
  path: string; // Ruta del hijo
  index?: boolean; // Indica si es la ruta principal (index)
  element: ReactElement; // Componente asociado
}

// Tipo para la ruta principal
interface RouteDef {
  path: string; // Ruta del padre
  element: ReactElement; // Componente principal
  errorElement?: ReactElement; // Componente en caso de error
  children?: RouteChild[]; // Rutas hijas
}

// Tipo completo para la definici√≥n del router
export type RouterDef = RouteDef[];