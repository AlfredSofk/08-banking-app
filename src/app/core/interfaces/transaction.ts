export interface ITransaction {
    id: string;
    amountTransaction: number;
    transactionCost: number;
    typeTransaction: string;
    timestamp: Date;
}

export interface IBankTransactionWithDrawFromATM {
    accountNumber :	string
    amount:	number
}


