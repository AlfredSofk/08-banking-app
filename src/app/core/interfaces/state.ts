import { IAuthState } from "./auth";
import { IAppState } from "./bankAccount";

export interface IContextApp {
    state : IStateAppContext
    dispatch : React.Dispatch<IActionApp>
}

export interface IStateAppContext extends IAppState {}

export interface IActionApp {
    type : string;
    payload : any;
}




export interface IContextAuth {
    state : IStateAuthContext
    dispatch : React.Dispatch<IActionAuth>
}

export interface IStateAuthContext extends IAuthState {}

export interface IActionAuth {
    type : string;
    payload : any;
}