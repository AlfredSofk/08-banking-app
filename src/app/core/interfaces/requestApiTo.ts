import { IDinError, IDinHeader } from "./requestToApi";

interface dinBodyLoginToken {
    token: string
}

interface BankTransactionDepositSucursal {
    accountNumberClient: string;
    amount: number;
}

interface IAccounts {
    number: string;
    amount: number;
}

interface IResDinBodyGetDataClient {
    username: string;
    rol: string | null;
    accounts: IAccounts[];
}

interface IResTransaction {
    id: string;
    amountTransaction: number;
    transactionCost: number;
    typeTransaction: string;
    timeStamp: Date
}
export interface ILoginToken {
    dinHeader: IDinHeader;
    dinBody: dinBodyLoginToken
    dinError: IDinError;
}

export interface ITransactionDeposit {
    dinHeader: IDinHeader;
    dinBody: BankTransactionDepositSucursal;
}

export interface IResBankTransactionWithDrawFromATM {
    dinHeader: IDinHeader;
    dinError: IDinError
}

export interface IResGetDataClient {
    dinHeader: IDinHeader;
    dinBody: IResDinBodyGetDataClient;
    dinError: IDinError;
}

export interface IResDataTransaction {
    dinHeader: IDinHeader;
    dinBody: IResTransaction;
    dinError: IDinError;
}


