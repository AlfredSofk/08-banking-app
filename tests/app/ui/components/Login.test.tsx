import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import LoginForm from '../../../../src/app/ui/components/Login';
import React from 'react';
import { vi } from 'vitest';
import { IAuthState } from '../../../../src/app/core/interfaces/auth';

const mockLogin = vi.fn();

const mockState : IAuthState = {    
    isAuthenticated: false,
    user: null,
    token: null,
    loading: false,
    error: null,
};


describe('Test para el componente Login',  () => {

    test('match snapshot', () => {
        const { asFragment } = render( 
            <LoginForm state={mockState} loginUser={mockLogin} />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test('Deberia mostrar los mensajes de error cuando se envía un formulario inválido', async() => {

        const { getByRole, findAllByRole } = render( <LoginForm loginUser={mockLogin} state={mockState} />);

        const buttonLogin = getByRole('buttonLogin');

        fireEvent.submit(buttonLogin);

        expect(await findAllByRole('alert')).toHaveLength(2);
        expect(mockLogin).not.toBeCalled()
    })

    test('Debe mostrar un mensaje de error cuando el usuario no es válido', async() => {
        const { getByRole, getByLabelText } = render( <LoginForm loginUser={mockLogin} state={mockState} />);

        const userMock : string = "=-%^&"

        fireEvent.input(getByRole("textbox", { name: /usuario/i }), {
            target: {
              value: userMock
            },
          })


        fireEvent.input(getByLabelText("Contraseña"), {
            target: {
              value: "123456789",
            },
        })

        fireEvent.submit(getByRole('buttonLogin'));
        expect(await screen.findAllByRole("alert")).toHaveLength(1)
        expect(mockLogin).not.toBeCalled()
        expect(getByRole("textbox", { name: /usuario/i })).toHaveValue(userMock)
        expect(getByLabelText("Contraseña")).toHaveValue("123456789")
    })

    test('Debe mostrar un mensaje de error cuando la contraseña no es válida', async() => {
        const { getByRole, getByLabelText } = render( <LoginForm loginUser={mockLogin} state={mockState} />);

        const userMock : string = "pablo"
        const passwordMock : string = ""

        fireEvent.input(getByRole("textbox", { name: /usuario/i }), {
            target: {
              value: userMock,
            },
          })

          fireEvent.input(getByLabelText("Contraseña"), {
            target: {
              value: passwordMock,
            },
          })

          fireEvent.submit(getByRole('buttonLogin'));
          expect(await screen.findAllByRole("alert")).toHaveLength(1)
          expect(mockLogin).not.toBeCalled()
          expect (getByRole("textbox", { name: /usuario/i })).toHaveValue(userMock)
          expect(getByLabelText("Contraseña")).toHaveValue(passwordMock)
    });

    test('No deberia mostrar un mensaje de error cuando los datos son validos', async() => {
        const { getByRole, getByLabelText, queryAllByRole } = render( <LoginForm loginUser={mockLogin} state={mockState} />);

        const userMock : string = "pablo"
        const passwordMock : string = "123456789"

        fireEvent.input(getByRole("textbox", { name: /usuario/i }), {
            target: {
              value: userMock,
            },
          })

          fireEvent.input(getByLabelText("Contraseña"), {
            target: {
              value: passwordMock,
            },
          })

          fireEvent.submit(getByRole('buttonLogin'));
        //   expect(await queryAllByRole("alert")).toHaveLength(0)

          await waitFor(() => expect(queryAllByRole("alert")).toHaveLength(0))
          expect(mockLogin).toBeCalledWith({ username: userMock, password: passwordMock })
          expect (getByRole("textbox", { name: /usuario/i })).toHaveValue(userMock)
          expect(getByLabelText("Contraseña")).toHaveValue(passwordMock)
    });
});

