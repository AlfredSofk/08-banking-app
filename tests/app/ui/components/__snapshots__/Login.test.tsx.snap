// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test para el componente Login > match snapshot 1`] = `
<DocumentFragment>
  <div
    class="login"
  >
    <section
      aria-labelledby="login-heading"
      class="login__container"
    >
      <h1
        class="login__title"
        id="login-heading"
      >
        Iniciar Sesión
      </h1>
      <form
        aria-describedby="login-form-desc"
        class="login-form"
      >
        <p
          class="login-form__desc"
          id="login-form-desc"
        >
          Introduce tus credenciales para acceder a tu cuenta.
        </p>
        <div
          class="login-form__group"
        >
          <label
            class="login-form__label"
            for="usuario"
          >
            Usuario
          </label>
          <input
            aria-describedby="usuario-desc"
            aria-invalid="false"
            aria-required="true"
            class="login-form__input"
            id="usuario"
            name="usuario"
            placeholder="usuarioEjemplo"
            type="usuario"
          />
          <span
            class="sr-only"
            id="usuario-desc"
          >
            Introduce un correo válido
          </span>
        </div>
        <div
          class="login-form__group"
        >
          <label
            class="login-form__label"
            for="password"
          >
            Contraseña
          </label>
          <input
            aria-describedby="password-desc"
            aria-invalid="false"
            aria-required="true"
            class="login-form__input"
            id="password"
            name="password"
            placeholder="••••••••"
            type="password"
          />
          <span
            class="sr-only"
            id="password-desc"
          >
            Introduce tu contraseña
          </span>
        </div>
        <button
          class="login-form__button"
          role="buttonLogin"
          type="submit"
        >
          Iniciar Sesión
        </button>
      </form>
      <p
        class="login__register"
      >
        ¿No tienes una cuenta?
      </p>
    </section>
  </div>
</DocumentFragment>
`;
